From f806acbd44b55b5f6de7ea16d3d84662cec6c98d Mon Sep 17 00:00:00 2001
From: Gilly1970 <gilroyscott@hotmail.com>
Date: Sun, 21 Sep 2025 22:48:08 +1000
Subject: [PATCH] 0129-mtk-mt76-mt7996-fix-kernel-6.104-EEPROM-0s

---
 mt7996/eeprom.c | 50 ++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 49 insertions(+), 1 deletion(-)

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 612ff3f..08f19d7 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -209,6 +209,44 @@ static bool mt7996_eeprom_variant_valid(struct mt7996_dev *dev, const u8 *def)
 	return true;
 }
 
+static void
+mt7996_eeprom_fixup_tx_power(struct mt7996_dev *dev, const u8 *def)
+{
+	u8 *eeprom = dev->mt76.eeprom.data;
+	int i;
+	bool zeros_detected = false;
+	bool sixg_overridden = false;
+
+	if (!eeprom[MT_EE_TX0_POWER_2G]) {
+		eeprom[MT_EE_TX0_POWER_2G] = def[MT_EE_TX0_POWER_2G];
+		zeros_detected = true;
+	}
+
+	for (i = MT_EE_TX0_POWER_5G; i < MT_EE_TX0_POWER_5G + 5; ++i) {
+		if (!eeprom[i]) {
+			eeprom[i] = def[i];
+			zeros_detected = true;
+		}
+	}
+
+	/*
+	 * Unconditionally use default firmware values for the 6GHz band.
+	 * This is a workaround for hardware with incorrect (but non-zero)
+	 * tx_power values in the EEPROM.
+	 */
+	for (i = MT_EE_TX0_POWER_6G; i < MT_EE_TX0_POWER_6G + 8; ++i) {
+		if (eeprom[i] != def[i]) {
+			eeprom[i] = def[i];
+			sixg_overridden = true;
+		}
+	}
+
+	if (zeros_detected)
+		dev_warn(dev->mt76.dev, "eeprom tx_power zeros detected, using defaults\n");
+	if (sixg_overridden)
+		dev_warn(dev->mt76.dev, "Overriding 6GHz eeprom tx_power with firmware defaults\n");
+}
+
 static int
 mt7996_eeprom_check_or_use_default(struct mt7996_dev *dev, bool use_default)
 {
@@ -262,6 +300,8 @@ mt7996_eeprom_load_bin(struct mt7996_dev *dev)
 	memcpy(eeprom, fw->data, MT7996_EEPROM_SIZE);
 	dev->flash_mode = true;
 
+	mt7996_eeprom_fixup_tx_power(dev, fw->data);
+
 out:
 	release_firmware(fw);
 	return ret;
@@ -367,7 +407,15 @@ static int mt7996_eeprom_load(struct mt7996_dev *dev)
 				goto out;
 			}
 		}
-	}
+
+		/* read tx_power values from fw */
+		u8 *eeprom = dev->mt76.eeprom.data;
+		if (!eeprom[MT_EE_TX0_POWER_2G] || !eeprom[MT_EE_TX0_POWER_5G] || !eeprom[MT_EE_TX0_POWER_6G] ) {
+			use_default = true;
+			goto out;
+		}	
+
+}
 
 out:
 	return mt7996_eeprom_check_or_use_default(dev, use_default);
-- 
2.43.0

